# Nextflow Configuration

This README documents the parameters used in the two nextflow configuration files used by Talos. The variables in these configuration files are initially set up to facilitate the demonstration test-run.

# annotation.config

[annotation.config](../nextflow/annotation.config)


| **Parameter**           | **Description**                                                                                                                                                                                                            |
|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `cohort`                | Name used to label this dataset throughout the output (directories and files)                                                                                                                                              |
| `cohort_output_dir`     | Output directory for cohort-specific Talos results, incorporates `cohort` in path                                                                                                                                          |
| `input_vcf_dir`         | Directory containing individual VCFs (e.g. *.vcf.gz)                                                                                                                                                                       |
| `merged_vcf`            | Path to a pre-merged multi-sample VCF file. A corresponding .tbi index is required. This file should contain `INFO.AC/AF/AN`. If present, this supersedes `input_vcf_dir`                                                  |
| `large_files`           | Directory containing large external resources (e.g. gnomAD, MANE, AlphaMissense). See [large_files README](../large_files/README.md) for instructions on how to source this content                                        |
| `ref_genome`            | Path in `large_files` to the reference genome FASTA file                                                                                                                                                                   |
| `gnomad_zip`            | Path in `large_files` to the echtvar annotation file from [Zenodo](https://zenodo.org/records/15222100)                                                                                                                    |
| `mane`                  | Path in `large_files` to the MANE transcript data                                                                                                                                                                          |
| `ensembl_gff`           | Path in `large_files` to the Ensembl GFF3 file                                                                                                                                                                             |
| `processed_annotations` | Directory for cohort-independent pre-processed annotation files. During a first run, these entries will be used as output paths. If accessible during subsequent runs these outputs will be reused, instead of regenerated |
| `mane_json`             | Path in `processed_annotations` to pre-processed MANE data                                                                                                                                                                 |
| `ensembl_bed`           | Path in `processed_annotations` to the Ensembl BED file, one line per gene                                                                                                                                                 |
| `ensembl_merged_bed`    | Path in `processed_annotations` to the Ensembl BED file, overlapping regions merged                                                                                                                                        |
| `alphamissense_tar`     | Path in `processed_annotations` to the AlphaMissense data, reformatted to be used in Hail annotation                                                                                                                       |
| `container`             | Docker image to use for the annotation workflow                                                                                                                                                                            |
| `docker.enabled`        | Parameter for Nextflow to enable usage of Docker                                                                                                                                                                           |

# talos.config

[talos.config](../nextflow/talos.config)

| **Parameter**           | **Description**                                                                                                                                                                                                                                                                                                             |
|-------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `cohort`                | Name used to label this dataset throughout the output (directories and files). This should be consistent with the same variable in `annotation.config`                                                                                                                                                                      |
| `input_dir`             | Directory for various input files. By default contents of this directory suitable for the test run, and should be replaced with real content. A different folder can be indicated at runtime using `--input_dir XXX`                                                                                                        |
| `pedigree`              | Path in `input_dir`, a `.ped` format file. Default expected path is `pedigree.ped` within `input_dir`                                                                                                                                                                                                                       |
| `runtime_config`        | Path in `input_dir`, the runtime configuration for Talos. See [Configuration.md](Configuration.md)                                                                                                                                                                                                                          |
| `output_dir`            | Output directory for the analysis, incorporates `cohort` in path                                                                                                                                                                                                                                                            |
| `phenopackets`          | Path in `output_dir` to a phenopackets file. If it doesn't exist at runtime, it will be generated. If `hpo_pedigree` exists it will be generated with embedded HPO terms, if only a plain `pedigree` is available it will be created without HPO terms                                                                      |
| `processed_annotations` | Directory for cohort-independent pre-processed annotation files. Most entries in this folder would have been generated during the `annotation.nf` workflow                                                                                                                                                                  |
| `parsed_mane`           | Annotation's `params.mane_json`                                                                                                                                                                                                                                                                                             |
| `panelapp`              | The Talos workflow downloads all relevant PanelApp data during a run, and caches it as a JSON file. This result is valid for a month (the PanelApp Australia update cadence). If PanelApp data was previously downloaded by a separate run, that JSON can be passed as an arguemnt to prevent re-querying for the same data |
| `large_files`           | Directory containing large external resources (e.g. gnomAD, MANE, AlphaMissense). See [large_files README](../large_files/README.md))                                                                                                                                                                                       |
| `hpo`                   | Path in `large_files` to the HPO ontology file in `.obo` form                                                                                                                                                                                                                                                               |
| `gen2phen`              | Path in `large_files` to the Geno~Phenotype databset in `.txt` form                                                                                                                                                                                                                                                         |
| `phenio_db`             | Path in `large_files` to the Monarch Phenotype DB in `.db.gz` form                                                                                                                                                                                                                                                          |
| `clinvar`               | Path in `large_files`, the latest monthly ClinvArbitration data from [Zenodo](https://zenodo.org/records/15896156)                                                                                                                                                                                                          |
| `container`             | Docker image to use for the Talos workflow                                                                                                                                                                                                                                                                                  |
| `docker.enabled`        | Parameter for Nextflow to enable usage of Docker                                                                                                                                                                                                                                                                            |
| `matrix_table`          | This is the output of the previous `annotation.nf` workflow. It has no default value, and the path to the MatrixTable should be passed at runtime using the `--matrix_table` command line argument                                                                                                                          |
| `ext_id_map`            | This is an optional file. If provided it can be a headerless 2-column CSV or TSV, or a JSON dictionary. In all cases it should map the Sample IDs in the variant data to an External ID. When provided, the HTML report will contain both the Sample ID and corresponding external ID                                       |
