# Nextflow Configuration

This README documents the parameters used in the nextflow configuration file. Some default values for configuration variables are initially set up to facilitate the demonstration test-run.

| **Parameter**           | **Description**                                                                                                                                                                                                                                                                                                                      |
|-------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `large_files`           | Directory containing large external resources (e.g. gnomAD, MANE, AlphaMissense). See [large_files README](../large_files/README.md) for instructions on how to source this content                                                                                                                                                  |
| `processed_annotations` | Output directory for all processed/reformatted data. Cohort/callset independent, to be used across multiple runs.                                                                                                                                                                                                                    |
| `cohort`                | Name used to label this dataset throughout the output (directories and files)                                                                                                                                                                                                                                                        |
| `cohort_output_dir`     | Output directory for cohort-specific Talos results.                                                                                                                                                                                                                                                                                  |
| `outputDir`             | Defaults to `processed_annotations`, required for the 'workflow outputs' NextFlow functionality. Override for Annotation/Talos workflow using `-output-dir <path>`                                                                                                                                                                   |
| `vcf`                   | Path to a pre-merged multi-sample VCF file. A corresponding .tbi index is required. This file should contain `INFO.AC/AF/AN`.                                                                                                                                                                                                        |
| `vcf_split_n`           | Combines with `vcf` - the input VCF is sharded into files containing `vcf_split_n` variant rows, and annotation operations are parallelised.                                                                                                                                                                                         |
| `shards`                | Optional, alternative to `vcf`. If used this should point to a directory containing multiple joint-call shards. The annotation pipeline will handle each shard in parallel.                                                                                                                                                          |
| `input_vcf_extension`   | Linked to `shards`. If a sharded VCF is provided, this argument is the file extension to glob for. By default this is `vcf.bgz`, but could be altered to `vcf.gz`.                                                                                                                                                                   |
| `pedigree`              | A `.ped` format file. Replace with real data                                                                                                                                                                                                                                                                                         |
| `alphamissense_tsv`     | Path in `large_files` to the AlphaMissense raw data, reformatted to be used in Hail annotation                                                                                                                                                                                                                                       |
| `alphamissense_zip`     | Path in `processed_annotations` to the reformatted AlphaMissense data, ready to be used by Echtvar for annotation                                                                                                                                                                                                                    |
| `ensembl_bed`           | Path in `processed_annotations` to the Ensembl BED file, one line per gene                                                                                                                                                                                                                                                           |
| `ensembl_merged_bed`    | Path in `processed_annotations` to the Ensembl BED file, overlapping regions merged                                                                                                                                                                                                                                                  |
| `mane_json`             | Path in `processed_annotations` to pre-processed MANE data                                                                                                                                                                                                                                                                           |
| `ensembl_gff`           | Path in `large_files` to the Ensembl GFF3 file                                                                                                                                                                                                                                                                                       |
| `gnomad_zip`            | Path in `large_files` to the echtvar annotation file from [Zenodo](https://zenodo.org/records/15222100)                                                                                                                                                                                                                              |
| `mane`                  | Path in `large_files` to the MANE transcript data                                                                                                                                                                                                                                                                                    |
| `ref_genome`            | Path in `large_files` to the reference genome FASTA file                                                                                                                                                                                                                                                                             |
| `runtime_config`        | Runtime configuration for Talos. See [Configuration.md](Configuration.md)                                                                                                                                                                                                                                                            |
| `hpo`                   | Path in `large_files` to the HPO ontology file in `.obo` form                                                                                                                                                                                                                                                                        |
| `gen2phen`              | Path in `large_files` to the Geno~Phenotype databset in `.txt` form                                                                                                                                                                                                                                                                  |
| `phenio_db`             | Path in `large_files` to the Monarch Phenotype DB in `.db.gz` form                                                                                                                                                                                                                                                                   |
| `ext_id_map`            | Optional. Headerless 2-column CSV or TSV, or a JSON dictionary. It maps Sample IDs in the variant data to an External ID. When provided, the HTML report will contain both the Sample ID and corresponding external ID                                                                                                               |
| `seqr_lookup`           | Optional. Headerless 2-column CSV or TSV, or a JSON dictionary. Maps Sample IDs in the variant data to a Seqr family ID. In conjunction with the [Configuration:Hyperlinks](https://github.com/populationgenomics/talos/blob/main/docs/Configuration.md#stage-createtaloshtml) config section, which provides the template Strings.  |
| `submission_summary`    | FTP URL for weekly ClinVar submission dump                                                                                                                                                                                                                                                                                           |
| `variant_summary`       | FTP URL for weekly ClinVar variant dump                                                                                                                                                                                                                                                                                              |
| `clinvar_blacklist`     | A string, containing quoted, space-delimited entries for all ClinVar submitter sites to blacklist (ignore).                                                                                                                                                                                                                          |
| `container`             | Docker image to use for the annotation workflow                                                                                                                                                                                                                                                                                      |
| `docker.enabled`        | Parameter for Nextflow to enable usage of Docker                                                                                                                                                                                                                                                                                     |
